File: backend/src/application/use-cases/search/TriggerManualScrapeUseCase.ts
import type { TriggerManualScrapeUseCasePort } from "../../ports/incoming/TriggerManualScrapeUseCasePort";
import type { SupermarketSearchPort } from "../../ports/outgoing/SupermarketSearchPort";
import type { IProduct } from "../../../interfaces/IProduct";
import { logger } from "../../../utils/logger";
import type { SearchResult, TriggerManualScrapeInput } from "./contracts";

export class TriggerManualScrapeUseCase implements TriggerManualScrapeUseCasePort {
	constructor(private readonly scrapers: SupermarketSearchPort[]) {}

	async execute(input: TriggerManualScrapeInput): Promise<SearchResult> {
		const query = input.query;
		const scrapedAt = new Date().toISOString();
		const warnings: string[] = [];

		logger.info(
			`[TriggerManualScrapeUseCase] Starting parallel scrapers for: "${query}"`,
		);
		const searchStartedAt = Date.now();

		const scraperTasks = this.scrapers.map(async (scraper) => {
			const scraperStartedAt = Date.now();
			try {
				const results = await scraper.search(query);
				logger.info(
					`[TriggerManualScrapeUseCase] ${scraper.name} completed: ${results.length} results in ${Date.now() - scraperStartedAt}ms`,
				);
				return results;
			} catch (error) {
				logger.error(
					`[TriggerManualScrapeUseCase] ${scraper.name} failed after ${Date.now() - scraperStartedAt}ms: ${error}`,
				);
				throw error;
			}
		});

		const settledResults = await Promise.allSettled(scraperTasks);
		logger.info(
			`[TriggerManualScrapeUseCase] All scrapers finished in ${Date.now() - searchStartedAt}ms.`,
		);

		const results: IProduct[] = [];
		settledResults.forEach((outcome, index) => {
			const scraperName = this.scrapers[index].name;
			if (outcome.status === "fulfilled") {
				results.push(...outcome.value);
				return;
			}

			const rejectionReason = this.formatRejectionReason(outcome.reason);
			logger.error(
				`[TriggerManualScrapeUseCase] ${scraperName} rejected: ${rejectionReason}`,
			);
			warnings.push(`${scraperName}: ${rejectionReason}`);
		});

		results.sort((left, right) => left.pricePerUnit - right.pricePerUnit);

		return {
			query,
			results,
			warnings,
			totalCount: results.length,
			source: "live",
			scrapedAt,
		};
	}

	private formatRejectionReason(reason: unknown): string {
		if (reason instanceof Error) return reason.message;
		if (typeof reason === "string") return reason;
		try {
			return JSON.stringify(reason);
		} catch {
			return String(reason);
		}
	}
}

File: backend-dotnet/src/api/Application/UseCases/Search/TriggerManualScrapeUseCase.cs
using Microsoft.Extensions.Logging;
using SharpMonads.Core;
using SupermarketPlanner.Api.Application.DTOs;
using SupermarketPlanner.Api.Application.Ports.Driven;
using SupermarketPlanner.Api.Application.Ports.Driving;
using SupermarketPlanner.Api.Domain;
using SupermarketPlanner.Api.Domain.Services;

namespace SupermarketPlanner.Api.Application.UseCases.Search;

public sealed class TriggerManualScrapeUseCase(
    IReadOnlyList<ISupermarketSearchPort> scrapers,
    ILogger<TriggerManualScrapeUseCase> logger) : ITriggerManualScrapeUseCase
{
    public async Task<Either<DomainError, SearchResult>> Invoke(
        TriggerManualScrapeInput input,
        CancellationToken cancellationToken)
    {
        logger.LogInformation(
            "[TriggerManualScrapeUseCase] Starting parallel scrapers for: \"{Query}\"",
            input.Query);

        var startedAt = DateTimeOffset.UtcNow;
        var scraperOutcomes = await Task.WhenAll(
            scrapers.Select(scraper => RunScraper(scraper, input, cancellationToken)));

        logger.LogInformation(
            "[TriggerManualScrapeUseCase] All scrapers finished in {ElapsedMs}ms.",
            (DateTimeOffset.UtcNow - startedAt).TotalMilliseconds);

        var results = (from outcome in scraperOutcomes
                       where outcome.Products is not null
                       from product in outcome.Products
                       orderby product.PricePerUnit
                       select product).ToList();

        var warnings = (from outcome in scraperOutcomes
                        where outcome.Products is null
                        select $"{outcome.ScraperName}: {outcome.Error ?? "Unknown error"}").ToList();

        return Either<DomainError, SearchResult>.FromRight(new SearchResult(
            Query: input.Query,
            Results: results,
            Warnings: warnings,
            TotalCount: results.Count,
            NextCursor: null,
            Source: "live",
            ScrapedAt: DateTimeOffset.UtcNow.ToString("o"),
            ActiveRefresh: RefreshDecision.None));
    }

    private async Task<(string ScraperName, IReadOnlyList<ProductDto>? Products, string? Error)> RunScraper(
        ISupermarketSearchPort scraper,
        TriggerManualScrapeInput input,
        CancellationToken cancellationToken)
    {
        var scraperStartedAt = DateTimeOffset.UtcNow;
        try
        {
            var searchResult = await scraper.Search(input.Query, input.PostalCode, cancellationToken);
            return searchResult.Match<(string, IReadOnlyList<ProductDto>?, string?)>(
                onLeft: error =>
                {
                    logger.LogError(
                        "[TriggerManualScrapeUseCase] {Name} failed in {ElapsedMs}ms: {Error}",
                        scraper.Name, (DateTimeOffset.UtcNow - scraperStartedAt).TotalMilliseconds, error.Message);
                    return (scraper.Name, null, error.Message);
                },
                onRight: products =>
                {
                    logger.LogInformation(
                        "[TriggerManualScrapeUseCase] {Name} completed: {Count} results in {ElapsedMs}ms",
                        scraper.Name, products.Count, (DateTimeOffset.UtcNow - scraperStartedAt).TotalMilliseconds);
                    return (scraper.Name, products, null);
                });
        }
        catch (Exception ex)
        {
            logger.LogError(
                "[TriggerManualScrapeUseCase] {Name} failed after {ElapsedMs}ms: {Error}",
                scraper.Name, (DateTimeOffset.UtcNow - scraperStartedAt).TotalMilliseconds, ex.Message);
            return (scraper.Name, null, ex.Message);
        }
    }
}
